### ISSUE TITLE
mapArgs is modified without locking

### ISSUE THREAD

**Issue by daira** [zcash/zcash#2132]
https://github.com/zcash/zcash/blob/master/src/rpcmining.cpp#L315

@laanwj wrote [here](https://github.com/bitcoin/bitcoin/pull/5957#r27958516):
> We should probably change this at some point too to not update the ``mapArgs``. Throughout the program we assume that ``mapArgs`` is read-only after initialization. This kind of r/w access would need proper locking around everything.

The problem was present in Bitcoin before we forked.

**Comment by laanwj** [comment]
See https://github.com/bitcoin/bitcoin/pull/9243

**Comment by nathan-at-least** [comment]
I left this in on the assumption the rebase from upstream is easy enough. Let's time box it.

**Comment by str4d** [comment]
Here's the chain of PRs that I've currently identified for merging (to reduce conflicts):

- https://github.com/bitcoin/bitcoin/pull/6077 (going to take some time, conflicts with changes we made)
- https://github.com/bitcoin/bitcoin/pull/6235
- https://github.com/bitcoin/bitcoin/pull/6961 (parts of, this one will be annoying)
- https://github.com/bitcoin/bitcoin/pull/8856
- https://github.com/bitcoin/bitcoin/pull/9243

I'll update the list as I find others.

**Comment by str4d** [comment]
Another parent PR is https://github.com/bitcoin/bitcoin/pull/6870 which itself should be merged after `-maxmempool` is added. As IIRC we have already discussed including that, I will do so in its own PR, so I will make a PR now with the earlier misc ones.

**Comment by nathan-at-least** [comment]
Bumped out of 1.0.9 for a few reasons: First I want to require more scrutiny of upstream PRs, and second because we're out of time for this issue.
